src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1010 @@
+
+
+-- | The standard pseudo-random number generator.
+newtype StdGen = StdGen { unStdGen :: SM.SMGen }
+  deriving newtype (RandomGen, NFData)
+  deriving stock   (Show)
+
+instance Eq StdGen where
+  StdGen x1 == StdGen x2 = SM.unseedSMGen x1 == SM.unseedSMGen x2
+
+instance RandomGen SM.SMGen where
+  next = SM.nextInt
+  genWord32 = SM.nextWord32
+  genWord64 = SM.nextWord64
+  split = SM.splitSMGen
+
+instance RandomGen SM32.SMGen where
+  next = SM32.nextInt
+  genWord32 = SM32.nextWord32
+  genWord64 = SM32.nextWord64
+  split = SM32.splitSMGen
+
+-- | Constructs a 'StdGen' deterministically.
+mkStdGen :: Int -> StdGen
+mkStdGen = StdGen . SM.mkSMGen . fromIntegral
+
+-- | The class of types for which a uniformly distributed value can be drawn
+-- from all possible values of the type.
+--
+-- @since 1.2
+class Uniform a where
+  -- | Generates a value uniformly distributed over all possible values of that
+  -- type.
+  --
+  -- @since 1.2
+  uniformM :: StatefulGen g m => g -> m a
+
+-- | The class of types for which a uniformly distributed value can be drawn
+-- from a range.
+--
+-- @since 1.2
+class UniformRange a where
+  -- | Generates a value uniformly distributed over the provided range, which
+  -- is interpreted as inclusive in the lower and upper bound.
+  --
+  -- *   @uniformRM (1 :: Int, 4 :: Int)@ generates values uniformly from the
+  --     set \(\{1,2,3,4\}\)
+  --
+  -- *   @uniformRM (1 :: Float, 4 :: Float)@ generates values uniformly from
+  --     the set \(\{x\;|\;1 \le x \le 4\}\)
+  --
+  -- The following law should hold to make the function always defined:
+  --
+  -- > uniformRM (a, b) = uniformRM (b, a)
+  --
+  -- @since 1.2
+  uniformRM :: StatefulGen g m => (a, a) -> g -> m a
+
+instance UniformRange Integer where
+  uniformRM = uniformIntegralM
+
+instance UniformRange Natural where
+  uniformRM = uniformIntegralM
+
+instance Uniform Int8 where
+  uniformM = fmap (fromIntegral :: Word8 -> Int8) . uniformWord8
+instance UniformRange Int8 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int8 -> Word8) fromIntegral
+
+instance Uniform Int16 where
+  uniformM = fmap (fromIntegral :: Word16 -> Int16) . uniformWord16
+instance UniformRange Int16 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int16 -> Word16) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Int32 where
+  uniformM = fmap (fromIntegral :: Word32 -> Int32) . uniformWord32
+instance UniformRange Int32 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int32 -> Word32) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Int64 where
+  uniformM = fmap (fromIntegral :: Word64 -> Int64) . uniformWord64
+instance UniformRange Int64 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int64 -> Word64) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Int where
+#if WORD_SIZE_IN_BITS < 64
+  uniformM = fmap (fromIntegral :: Word32 -> Int) . uniformWord32
```

Comment:
You can avoid CPP here and below using 
```haskell
uniformM = if finiteBitSize (0 :: Word) == 64 then ... else ...
```
GHC is clever enough to optimise such condition away.

Reply:
I am always careful about what ghc can and can not optimize away, have been burned too many times ;) Honestly, despite that it looks ugly, I'd rather follow the same approach taken in `base` for this, which is using `CPP`

Reply:
I used `finiteBitSize (0 :: Word) == 64` several times in performance critical parts and never had an issue with it. Getting rid of `CPP` makes tooling work more smoothly, which is a pretty sensible benefit IMO.

Reply:
I'll check it out and if benchmarks show no change I'll update, cause I do fancy the idea of getting rid of CPP